octoprint_printhistory/parser.py

```diff
@@ -0,0 +1,294 @@
+# coding=utf-8
+import re
+import os
+import io
+import unittest
+import ConfigParser
+
+
+VERSION_REGEX = re.compile(r"(\d+)?\.(\d+)?\.?(\*|\d+)")
+
+
+class UniversalParser():
+
+    def __init__(self, file_path):
+        self.parser_factory = None
+        self.name = None
+        self.version = None
+        self.file = open(file_path, "r")
+        for parser in [CuraParser(), Slic3rParser(), Simplify3DParser()]:
+            if parser.detect(self.file):
+                self.parser_factory = parser
+                self.name = parser.name
+                self.version = parser.version
+
+    def parse(self):
+        parameters = self.parser_factory.parse(self.file)
+        parameters.update({"slicer_name": self.name})
+        parameters.update({"slicer_version": self.version})
+        return parameters
+
+
+class BaseParser():
+
+    def parse(self, file):
+        parameters = {}
+        parameters.update(self.parse_header(file))
+        parameters.update(self.parse_bottom(file))
+        file.close()
+        return parameters
+
+
+class CuraParser(BaseParser):
+
+    def __init__(self):
+        self.name = "cura"
+        self.version = None
+
+    def detect(self, file):
+        detected = False
+        # on the third line
+        # ;Generated with Cura_SteamEngine 2.3.1
+        for _ in range(3):
+            line = file.readline()
+            if re.search(r"Cura_SteamEngine", line):
+                detected = True
+                self.version = VERSION_REGEX.search(line).group(0)
+        file.seek(0)
+        return detected
+
+    def parse_header(self, file):
+        parameters = {}
+        for i in range(15):
+            line = file.readline()
+            if line.startswith(";"):
+                line = line.replace(";", "")
+                if line.startswith("TIME") or line.startswith("LAYER_COUNT"):
+                    splitted = line.split(":", 1)
+                    parameters.update({splitted[0]: splitted[1].strip()})
+        file.seek(0)
+        return parameters
+
+    def parse_bottom(self, file):
+        parameters = {}
+        settings_reversed = []
+        settings = []
+        for line in reverse_readline(file):
+            if line.startswith(";SETTING_3"):
+                line = line.replace(";SETTING_3", "")
+                settings_reversed.append(line)
+            else:
+                break
+        [settings.append(x.strip()) for x in reversed(settings_reversed)]
+        settings = "".join(settings).replace("\\\\n", "\n")
+        settings = settings.replace('{"global_quality": "', "")
+        settings = settings.replace('"}', "")
+        config = ConfigParser.RawConfigParser(allow_no_value=True)
+        config.readfp(io.BytesIO(settings))
+        try:
+            for section in ["values", "metadata"]:
+                for option in config.options(section):
+                    parameters.update({option: config.get(section, option)})
+        except:
```

Comment:
never do naked `except:` clause, this is a straight antipattern even by PEP8 (http://stackoverflow.com/a/14797463/699864). Always specify a class or a tuple of classes.